src/diffusers/models/unet_motion_blocks.py

```diff
@@ -0,0 +1,1234 @@
+        hidden_states = hidden_states.to(query.dtype)
+
+        # linear proj
+        hidden_states = attn.to_out[0](hidden_states, scale=scale)
+        # dropout
+        hidden_states = attn.to_out[1](hidden_states)
+
+        if input_ndim == 4:
+            hidden_states = hidden_states.transpose(-1, -2).reshape(batch_size, channel, height, width)
+
+        if attn.residual_connection:
+            hidden_states = hidden_states + residual
+
+        hidden_states = hidden_states / attn.rescale_output_factor
+
+        return hidden_states
+
+
+class MotionBlock(nn.Module):
+    def __init__(
+        self,
+        in_channels,
+        norm_num_groups=32,
+        cross_attention_dim=None,
+        activation_fn="geglu",
+        attention_bias=False,
+        num_attention_heads=8,
+        max_seq_length=24,
+    ) -> None:
+        super().__init__()
+
+        self.temporal_transformer = TransformerTemporalModel(
+            in_channels=in_channels,
+            norm_num_groups=norm_num_groups,
+            cross_attention_dim=cross_attention_dim,
+            attention_head_dim=in_channels // num_attention_heads,
+            activation_fn=activation_fn,
+            attention_bias=attention_bias,
+            num_attention_heads=num_attention_heads,
+        )
+        processor_cls = MotionAttnProcessor2_0 if is_torch_version(">=", "2.0.0") else MotionAttnProcessor
+
+        for block in self.temporal_transformer.transformer_blocks:
+            block.attn1.set_processor(processor_cls(in_channels=in_channels, max_seq_length=max_seq_length))
+            block.attn2.set_processor(processor_cls(in_channels=in_channels, max_seq_length=max_seq_length))
+
+    def forward(self, hidden_states, encoder_hidden_states=None, num_frames=1):
+        hidden_states = self.temporal_transformer(
+            hidden_states, encoder_hidden_states=encoder_hidden_states, num_frames=num_frames
+        ).sample
+
+        return hidden_states
+
+
+class MotionModules(nn.Module):
+    def __init__(
+        self,
+        in_channels,
+        layers_per_block=2,
+        num_attention_heads=8,
+        attention_bias=False,
+        cross_attention_dim=None,
+        activation_fn="geglu",
+        norm_num_groups=32,
+        max_seq_length=32,
+    ):
+        super().__init__()
+        self.motion_modules = nn.ModuleList([])
+
+        for i in range(layers_per_block):
+            self.motion_modules.append(
+                MotionBlock(
+                    in_channels=in_channels,
+                    norm_num_groups=norm_num_groups,
+                    cross_attention_dim=cross_attention_dim,
+                    activation_fn=activation_fn,
+                    attention_bias=attention_bias,
+                    num_attention_heads=num_attention_heads,
+                    max_seq_length=max_seq_length,
+                )
+            )
+
+
+class MotionAdapter(ModelMixin, ConfigMixin):
```

Comment:
This should be in its own file. Because it's very special I'd probably just add it to the `animatediff` folder